### ISSUE TITLE
Passing dict instance produced by dict.fromkeys inline to str.maketrans results in type var confusion

### ISSUE THREAD

**Issue by layday** [microsoft/pyright#5065]
**Describe the bug**

The following error is produced when a dict is constructed from keys and passed inline as an argument to `str.maketrans`:

```
Argument of type "dict[str | int, _T@maketrans | str]" cannot be assigned to parameter "__x" of type "dict[int, _T@maketrans] | dict[str, _T@maketrans] | dict[str | int, _T@maketrans]" in function "maketrans"
  Type "dict[str | int, _T@maketrans | str]" cannot be assigned to type "dict[int, _T@maketrans] | dict[str, _T@maketrans] | dict[str | int, _T@maketrans]"
    "dict[str | int, _T@maketrans | str]" is incompatible with "dict[int, _T@maketrans]"
      TypeVar "_KT@dict" is invariant
        Type "str | int" cannot be assigned to type "int"
          "str" is incompatible with "int"
      TypeVar "_VT@dict" is invariant
        Type "_T@maketrans | Literal['b']" cannot be assigned to type "_T@maketrans | str"
    "dict[str | int, _T@maketrans | str]" is incompatible with "dict[str, _T@maketrans]"
```

Assigning the dict to an intermediate variable does not result in an error.

**To Reproduce**
Type check the snippet below.

**Expected behavior**
That no error would be produced.

**Code**
```python
# pyright: strict

a = dict.fromkeys('a', 'b')
b = str.maketrans(a)  # Ok
c = str.maketrans(dict.fromkeys('a', 'b'))  # Type error

```

**VS Code extension or command-line**
Either, v1.1.306

**Comment by erictraut** [comment]
Here's a test case that exhibits the problem without relying on complex builtin types with overloads.

```python
from __future__ import annotations
from typing import Generic, Iterable, TypeVar

_KT = TypeVar("_KT")
_VT = TypeVar("_VT")
_S = TypeVar("_S")
_T = TypeVar("_T")


class A(Generic[_KT, _VT]):
    @classmethod
    def method1(cls, i: Iterable[_T], v: _S) -> A[_T, _S]:
        ...

def func1(__x: A[str, _T] | A[str | int, _T]) -> A[int, _T]:
    ...

v1 = func1(A.method1("a", "b"))
```

The problem is related to the heuristics that pyright uses when performing bidirectional type inference. It gets complicated when the "expected type" is a union, as it is in this case (`A[str, _T] | A[str | int, _T]`). And it's even more complicated when multiple subtypes within the union could satisfy the return type of the inner call (`A[_T, _S]`).

I'll need to give this one more thought. I may need to resolve it as "won't fix" if the potential solutions break other cases or impose too high of a performance cost.

**Comment by erictraut** [comment]
This will be addressed in the next release.

**Comment by erictraut** [comment]
Unfortunately, I needed to back out this fix for now because it resulted in a regression that I need to investigate further.

**Comment by erictraut** [comment]
I've found a fix that appears to be sound (i.e. doesn't regress other cases). This will be included in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.310, which I just published. It will also be included in a future release of pylance.

**Comment by layday** [comment]
Thank you.

Sent with [Proton Mail](https://proton.me/) secure email.

------- Original Message -------
